File: test_notification_debug.js
// Script de test pour déboguer les notifications quotidiennes
// Usage: node test_notification_debug.js

console.log('🔍 DEBUG: Système de notifications quotidiennes DailyGrowth');
console.log('=' .repeat(60));

// 1. Vérifications temporelles
const now = new Date();
const utcTime = now.toISOString();
const localTime = now.toLocaleString('fr-FR', { timeZone: 'Europe/Paris' });

console.log('\n📅 INFORMATIONS TEMPORELLES:');
console.log(`Heure UTC: ${utcTime}`);
console.log(`Heure locale (Paris): ${localTime}`);
console.log(`Heure UTC: ${now.getUTCHours()}:${now.getUTCMinutes().toString().padStart(2, '0')}`);

// 2. Test de la logique de notification
const targetHour = 9; // 9h00 paramétré
const targetMinute = 0;
const currentHour = now.getUTCHours();
const currentMinute = now.getUTCMinutes();

const targetMinutes = targetHour * 60 + targetMinute;
const currentMinutes = currentHour * 60 + currentMinute;
const timeDiff = Math.abs(currentMinutes - targetMinutes);
const shouldSendNow = timeDiff <= 15 || (currentHour === targetHour && Math.abs(currentMinute - targetMinute) <= 15);

console.log('\n🎯 LOGIQUE DE NOTIFICATION:');
console.log(`Heure cible: ${targetHour}:${targetMinute.toString().padStart(2, '0')} UTC`);
console.log(`Heure actuelle: ${currentHour}:${currentMinute.toString().padStart(2, '0')} UTC`);
console.log(`Différence en minutes: ${timeDiff}`);
console.log(`Devrait envoyer maintenant: ${shouldSendNow ? 'OUI' : 'NON'}`);

// 3. Points à vérifier
console.log('\n✅ CHECKLIST DEBUG:');
console.log('1. [ ] Cron job configuré et actif dans Supabase');
console.log('2. [ ] Variables d\'environnement FIREBASE_SERVER_KEY configurée');
console.log('3. [ ] Utilisateur expertiaen5min@gmail.com existe avec:');
console.log('   - notifications_enabled = true');
console.log('   - reminder_notifications_enabled = true');
console.log('   - fcm_token renseigné');
console.log('   - notification_time = "09:00:00"');
console.log('4. [ ] Token FCM valide et non expiré');
console.log('5. [ ] Fonction Edge accessible et fonctionnelle');

console.log('\n🧪 PROCHAINES ACTIONS:');
console.log('1. Tester manuellement la fonction send-daily-notifications');
console.log('2. Vérifier les logs Supabase');
console.log('3. Contrôler la base de données utilisateur');
console.log('4. Tester l\'envoi de push notification direct');

File: test_notification_debug.dart
import 'dart:io';

void main() async {
  print('🔧 DEBUG SYSTÈME NOTIFICATIONS - expertiaen5min@gmail.com');
  print('=' * 60);
  
  // Simuler les données de l'utilisateur depuis la base
  final userId = '550e8400-e29b-41d4-a716-446655440001';
  final userEmail = 'expertiaen5min@gmail.com';
  final notificationTime = '14:30:00';
  final notificationsEnabled = true;
  
  print('👤 UTILISATEUR:');
  print('ID: $userId');
  print('Email: $userEmail');
  print('Heure notification: $notificationTime');
  print('Notifications activées: $notificationsEnabled');
  
  // Simuler la logique de NotificationService.updateNotificationSettings()
  print('\n🔧 SIMULATION updateNotificationSettings():');
  
  // 1. Calcul de l'offset timezone (ligne 357 du code)
  final timezoneOffsetMinutes = DateTime.now().timeZoneOffset.inMinutes;
  print('Offset timezone calculé: $timezoneOffsetMinutes minutes (${timezoneOffsetMinutes / 60}h)');
  
  // 2. Simulation de la mise à jour en base
  final updateData = {
    'notification_time': notificationTime,
    'notifications_enabled': notificationsEnabled,
    'reminder_notifications_enabled': false,
    'notification_timezone_offset_minutes': timezoneOffsetMinutes,
  };
  
  print('Données à sauvegarder: $updateData');
  
  // 3. Simulation de scheduleDailyNotification()
  print('\n📅 SIMULATION scheduleDailyNotification():');
  
  final timeParts = notificationTime.split(':');
  final hour = int.parse(timeParts[0]);
  final minute = int.parse(timeParts[1]);
  
  print('Heure parsée: ${hour}h${minute.toString().padLeft(2, '0')}');
  
  // 4. Simulation de _nextInstanceOfTime() (ligne 161-169)
  final now = DateTime.now();
  DateTime scheduledDate = DateTime(now.year, now.month, now.day, hour, minute);
  
  if (scheduledDate.isBefore(now)) {
    scheduledDate = scheduledDate.add(const Duration(days: 1));
    print('Heure passée → reportée à demain');
  } else {
    print('Heure future → programmée aujourd\'hui');
  }
  
  final timeUntil = scheduledDate.difference(now);
  
  print('Notification programmée pour: $scheduledDate');
  print('Dans: ${timeUntil.inHours}h ${timeUntil.inMinutes % 60}min');
  
  // 5. Vérification des problèmes potentiels
  print('\n🔍 DIAGNOSTIC PROBLÈMES POTENTIELS:');
  
  // Problème 1: Migration timezone manquante
  print('1. Migration timezone:');
  print('   - notification_timezone_offset_minutes calculé: ✅');
  print('   - Colonne existe en base: ❓ (à vérifier)');
  
  // Problème 2: Timezone hardcodé Europe/Paris
  print('2. Timezone hardcodé:');
  print('   - Code force Europe/Paris: ✅ (OK pour France)');
  print('   - Système local en France: ✅');
  
  // Problème 3: Différence Web vs Mobile
  print('3. Plateforme:');
  print('   - Web: Service Worker + heure locale navigateur');
  print('   - Mobile: flutter_local_notifications + timezone Paris');
  print('   - Cohérence: ✅ (si utilisateur en France)');
  
  // Problème 4: Gestion des erreurs
  print('4. Gestion erreurs:');
  print('   - Try/catch présent: ✅');
  print('   - Fallback UTC: ✅');
  print('   - Debug logs: ✅');
  
  // Test de conversion UTC
  print('\n🌍 TEST CONVERSION UTC:');
  final utcTime = scheduledDate.toUtc();
  print('Heure locale: $scheduledDate');
  print('Heure UTC: $utcTime');
  print('Différence: ${scheduledDate.difference(utcTime).inHours}h');
  
  // Recommandations
  print('\n💡 RECOMMANDATIONS:');
  if (timeUntil.inMinutes < 60) {
    print('🚨 URGENT: Notification dans ${timeUntil.inMinutes} minutes !');
    print('   → Tester immédiatement pour vérifier le fonctionnement');
  } else {
    print('📅 Notification dans ${timeUntil.inHours}h${timeUntil.inMinutes % 60}min');
    print('   → Programmer un test à 14:25 pour vérifier');
  }
  
  print('\n✅ SYSTÈME SEMBLE CORRECT POUR LA FRANCE');
  print('   → Problème potentiel: Migration timezone non appliquée en base');
  print('   → Solution: Appliquer la migration ou ignorer si France uniquement');
}
